runtime/src/iree-amd-aie/driver/xrt/direct_allocator.cc

```diff
@@ -0,0 +1,269 @@
+      compatibility |= IREE_HAL_BUFFER_COMPATIBILITY_QUEUE_DISPATCH;
+    }
+  }
+
+  // We are now optimal.
+  params->type &= ~IREE_HAL_MEMORY_TYPE_OPTIMAL;
+
+  // Guard against the corner case where the requested buffer size is 0. The
+  // application is unlikely to do anything when requesting a 0-byte buffer; but
+  // it can happen in real world use cases. So we should at least not crash.
+  if (*allocation_size == 0) *allocation_size = 4;
+
+  // Align allocation sizes to 4 bytes so shaders operating on 32 bit types can
+  // act safely even on buffer ranges that are not naturally aligned.
+  *allocation_size = iree_host_align(*allocation_size, 4);
+
+  return compatibility;
+}
+
+static iree_status_t iree_hal_xrt_allocator_allocate_buffer(
+    iree_hal_allocator_t* IREE_RESTRICT base_allocator,
+    const iree_hal_buffer_params_t* IREE_RESTRICT params,
+    iree_device_size_t allocation_size,
+    iree_hal_buffer_t** IREE_RESTRICT out_buffer) {
+  iree_hal_xrt_allocator_t* allocator =
+      iree_hal_xrt_allocator_cast(base_allocator);
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, allocation_size);
+
+  // Coerce options into those required by the current device.
+  iree_hal_buffer_params_t compat_params = *params;
+  if (!iree_all_bits_set(iree_hal_xrt_allocator_query_buffer_compatibility(
+                             base_allocator, &compat_params, &allocation_size),
+                         IREE_HAL_BUFFER_COMPATIBILITY_ALLOCATABLE)) {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "allocator cannot allocate a buffer with the given parameters");
+  }
+
+  iree_status_t status = iree_ok_status();
+  // comment from XRT
+  // Group id is the memory bank index where a global buffer
+  // can be allocated for use with this kernel.   If the kernel
+  // contains imcompatible compute units, then these are
+  // filtered out from a run object when the arguments are set.
+  // This filtering implies that the group id returned by this
+  // function may not necessarily be compatible with an existing
+  // filtered run object, but it is guaranteed to be compatible
+  // with a new 'fresh' run object.
+
+  // It seems host side bank is bank 8 so thats what we will set the
+  // group id as
+
+  int group_id = 8;
+
+  xrt::bo xrt_buffer = xrt::bo(allocator->device, allocation_size,
+                               XRT_BO_FLAGS_HOST_ONLY, group_id);
+
+  IREE_TRACE_ZONE_END(z0);
+  if (!xrt_buffer) {
+    status = iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED,
+                              "unable to allocate buffer");
+  }
+
+  iree_hal_buffer_t* buffer = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_xrt_buffer_wrap(
+        xrt_buffer, base_allocator, compat_params.type, compat_params.access,
+        compat_params.usage, allocation_size, /*byte_offset=*/0,
+        /*byte_length=*/allocation_size,
+        iree_hal_buffer_release_callback_null(), &buffer);  // +1
```

Comment:
Again, we should drop those `+1` or `-1` comments--they are only relevant to metal.